scripts/igmp_query_injector.py

```diff
@@ -0,0 +1,458 @@
+#!/usr/bin/env python
+import argparse
+import os
+import socket
+import subprocess
+from abc import ABCMeta, abstractmethod
+from contextlib import contextmanager
+from select import EPOLLIN, epoll
+import logging
+import signal
+import resource
+import xcp.logger as log
+import time
+import sys
+import re
+
+__DEBUG = False
+if __DEBUG:
+    logging_lvl = logging.DEBUG
+else:
+    logging_lvl = logging.INFO
+
+
+class Singleton(type):
+    _instances = {}
+
+    def __call__(cls, *args, **kwargs):
+        if cls not in cls._instances:
+            cls._instances[cls] = super(Singleton, cls).__call__(*args, **kwargs)
+        return cls._instances[cls]
+
+
+class SubprocessAutoClean(object):
+    __metaclass__ = Singleton
+    """Ensure all subprocess are terminated when the main process exits
+    """
+    CATCH_SIGNALS = [signal.SIGTERM, signal.SIGINT]
+
+    def __init__(self):
+        self.subprocess_lst = []
+
+    def add_subprocess(self, p):
+        self.subprocess_lst.append(p)
+
+    def del_subprocess(self, p):
+        self.subprocess_lst.remove(p)
+
+    def kill_all_subprocess(self):
+        if not self.subprocess_lst:
+            return
+
+        log.info('kill all sub process')
+        for p in self.subprocess_lst:
+            p.terminate()
+
+        self.subprocess_lst = []
+
+    def __enter__(self):
+        self._set_signal_handler(self.CATCH_SIGNALS, self._kill_all_subprocess_and_exit)
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        self.kill_all_subprocess()
+
+    def _set_signal_handler(self, signals, f):
+        for s in signals:
+            signal.signal(s, f)
+
+    def _kill_all_subprocess_and_exit(self, signum, frame):
+        log.critical('Catch signal %d, kill all subprocess' % signum)
+        self.kill_all_subprocess()
+        sys.exit(ERRNO.SIGNAL_CATCH)
+
+
+class ERRNO(object):
+    SUCC = 0
+    SIGNAL_CATCH = 1
+
+
+class XenstoreUtil(object):
+    """Utilities for access xenstore
+    """
+    @staticmethod
+    def read(path):
+        return subprocess.check_output(['xenstore-read', path]).strip()
+
+    @staticmethod
+    def vif_mac_path(domid, vifid):
+        return '/local/domain/%d/device/vif/%d/mac' % (domid, vifid)
+
+    @staticmethod
+    def vif_state_path(domid, vifid):
+        return '/local/domain/%d/device/vif/%d/state' % (domid, vifid)
+
+
+class XenstoreWatcher(object):
+    """Tool for watching xenstore
+    """
+    def __init__(self, watch_path_list):
+        self._watch_path_list = watch_path_list
+        self.p = None
+
+    def start(self):
+        self.p = subprocess.Popen(['xenstore-watch'] + self._watch_path_list, stdout=subprocess.PIPE)
```

Comment:
There is a 3rd party package https://pypi.python.org/pypi/pyxs, but is that a good idea introduce it into dom0?